src/v/cluster/partition.cc

```diff
@@ -1213,19 +1213,52 @@ partition::get_cloud_storage_manifest_view() {
 ss::future<result<model::offset, std::error_code>>
 partition::sync_kafka_start_offset_override(
   model::timeout_clock::duration timeout) {
-    if (_log_eviction_stm && !is_read_replica_mode_enabled()) {
+    if (is_read_replica_mode_enabled()) {
+        auto term = _raft->term();
+        if (!co_await _archival_meta_stm->sync(timeout)) {
+            if (term != _raft->term()) {
+                co_return errc::not_leader;
+            } else {
+                co_return errc::timeout;
+            }
+        }
+        auto start_kafka_offset
+          = _archival_meta_stm->manifest().get_start_kafka_offset_override();
+
+        co_return kafka::offset_cast(start_kafka_offset);
+    }
+
+    if (_log_eviction_stm) {
         auto offset_res = co_await _log_eviction_stm
                             ->sync_kafka_start_offset_override(timeout);
         if (offset_res.has_failure()) {
             co_return offset_res.as_failure();
         }
         if (offset_res.value() != kafka::offset{}) {
-            co_return offset_res.value();
+            co_return kafka::offset_cast(offset_res.value());
         }
+
         // If a start override is no longer in the offset translator state,
         // it may have been uploaded and persisted in the manifest.
     }
-    if (_archival_meta_stm) {
+
+    if (!_archival_meta_stm) {
+        co_return model::offset{};
+    }
+
+    // There are a few cases in which the log_eviction_stm will return a raft
+    // offset of `model::offset{}` for the start offset override. And a case
+    // where we won't be able to translate the raft offset it does return;
+    // - The topic was remotely recovered.
+    // - A start offset override was never set.
+    // - The raft offset is no longer on local disk.
+    //
+    // In all cases we'll need to sync the archival stm once to determine if a
+    // start offset override has been set. And if so what it was set to.
+    //
+    // Once it has been sync'd we can cache the override value as it won't
+    // change until `_log_eviction_stm->cached_kafka_start_offset()` does.
+    if (!_cached_kafka_start_override || !_log_eviction_stm) {
         auto term = _raft->term();
         if (!co_await _archival_meta_stm->sync(timeout)) {
```

Comment:
nit: may be clearer as some

```
if (_cached_kafka_start_override.has_value() && _log_eviction_stm) {
    return *_cached_kafka_start_override;
}
auto term = _raft->term();
...
```

Reply:
Actually I'm thinking it might make sense for us to early return very early, if `!_log_eviction_stm`, in which case we can omit these other checks for `_log_eviction_stm`.

It's probably worth checking if `_log_eviction_stm` can ever be removed -- I don't think it can, but maybe someone from the replication team will have a clearer understanding, since that code has changed in the last few months

Reply:
Currently there are a couple ntps that don't have a `_log_eviction_stm`. The controller, the consumer offsets, tx manager... Luckily none of these are ktps and none are consumable in the normal kafka manner. So for any partitions we're concerned about here the log eviction stm will always exist. I think I should still check for its existence though as I think it makes things clearer.

Reply:
>  Luckily none of these are ktps and none are consumable in the normal kafka manner

__consumer_offets is a kafka partition and AFAIK is consumable in the normal kafka way? IIRC this was a change we made a couple of years ago: it used to be in the Redpanda namespace and hence invisible to Kafka for for compatibility we moved it into `kafka` namespace with the same name as Apache Kafka so it would work in the same way for anyone who was accessing it directly (Enterprise team would probably have the most context here).

Reply:
Good point, then lets assume there are kafka consumable partitions without the log eviction stm(like in consumer offsets as you've mentioned). Then there are two paths the function can take;

Either `!_archival_meta_stm` in which case it returns `model::offset{}` which is correct since there can be no overridden offset without the log eviction stm.
Or `_archival_meta_stm` in which case it'll sync the archival stm once then return `_archival_meta_stm->manifest().get_start_kafka_offset_override()` for every call afterwards. The result of this function should be `model::offset{}` as well since we shouldn't technically be able to apply delete record batches to partitions without the log eviction stm

Either way the function does return the correct value.

Reply:
SG.

@andrwng any other concerns on this thread of discussion?

Reply:
Yea this looks good. Thanks for the ping Travis, and Brandon for updating!